Prevent auto-updating premium plugin with incompatible free version

triple0t opened the pull request:
## Description

This PR adds version compatibility checking to prevent the MailPoet Premium plugin from auto-updating when the installed free plugin version is incompatible, avoiding potential version mismatches and compatibility issues.

## Why

Currently, the premium plugin update mechanism doesn't verify version compatibility with the free plugin before showing update notices. This can lead to scenarios where:

- The premium plugin auto-updates to a version that requires a newer free plugin version than what's installed
- Users end up with incompatible premium/free plugin combinations
- Plugin functionality may break due to version mismatches
- Users experience errors or unexpected behavior

This issue is particularly problematic when:
- WordPress auto-updates are enabled
- The free plugin hasn't been updated yet (e.g., pending WordPress.org review)
- The premium plugin releases ahead of the free plugin

Fixes: STOMAIL-7527

| Before | After |
| ------ | ----- |
|    <img width="1578" height="300" alt="Screenshot 2025-11-14 at 3 52 22 PM" src="https://github.com/user-attachments/assets/47dd1d9a-674e-41d3-a8aa-0260391c3260" />  |     <img width="1512" height="254" alt="Screenshot 2025-11-14 at 3 54 06 PM" src="https://github.com/user-attachments/assets/fda697cb-ed1e-47c3-9ec3-58acdc5611bc" /> |

## Code review notes

### Version Compatibility Logic

The compatibility check works as follows:

1. **Extract major.minor versions**:
   - Premium `5.17.0` → `5.17`
   - Free `5.17.5` → `5.17`

2. **Compare versions**:
   - Compatible: Free version `>=` required premium version (e.g., `5.17` >= `5.17`)
   - Incompatible: Free version `<` required premium version (e.g., `5.16` < `5.17`)

3. **Compatibility examples**:
   - ✅ Premium `5.17.0` + Free `5.17.0` → Compatible
   - ✅ Premium `5.17.0` + Free `5.17.5` → Compatible
   - ✅ Premium `5.16.0` + Free `5.17.0` → Compatible
   - ❌ Premium `5.18.0` + Free `5.17.0` → Incompatible
   - ❌ Premium `6.0.0` + Free `5.17.0` → Incompatible



## QA notes

Before testing, ensure both the Free and Premium plugin versions are on 5.17.0 (the most recent version)

* Perform testing on the Updates page `/wp-admin/update-core.php` 
* Before checking out this branch, force a Premium plugin version update. You can do that by adding `$latestVersion->new_version = '5.18.0';` to this line https://github.com/mailpoet/mailpoet/blob/d3720e8d1ddb5d3c1a720742c8c1c0f59f84c7e2/mailpoet/lib/Config/Updater.php#L46
* Visit the update page. Force reload if necessary. Notice the website prompt for a MailPoet premium plugin update
* Checkout this PR
* Reload the update page.  
* Notice the MailPoet premium plugin update is gone.

Click here to force reload
<img width="581" height="192" alt="Screenshot 2025-11-14 at 3 52 37 PM" src="https://github.com/user-attachments/assets/143e1e52-6dc5-434c-a348-f62d588733e4" />


## Linked PRs

_N/A_

## Linked tickets

STOMAIL-7527

## After-merge notes

_N/A_

## Tasks

- [x] I have added a changelog entry (`./do changelog:add --type=<type> --description=<description>`)
- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:stomail-7527-prevent-auto-updating-the-premium-plugin-with-version)

_The latest successful build from `stomail-7527-prevent-auto-updating-the-premium-plugin-with-version` will be used. If none is available, the link won't work._

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Prevents incompatible premium plugin updates and suppresses update notices for versions that would conflict with the installed free plugin, avoiding failed or partial updates.

* **Documentation**
  * Added a changelog entry describing the improved update behavior.

* **Tests**
  * Expanded test coverage for update-check and version-compatibility scenarios to ensure correct notice and update behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

triple0t committed:
Prevent auto-updating the premium plugin with version mismatch

This ensures that the premium plugin can not be updated with incompatible free plugin version

triple0t committed:
Add tests and changelog files

triple0t committed:
Coder rabbit pr feedback

NeosinneR commented on mailpoet/lib/Config/Updater.php:
```diff
@@ -56,4 +74,31 @@ public function getLatestVersion() {
+    $requiredMainVersion = end($matches);
+
+    // Extract current main plugin minor version
+    preg_match('/^\d\.\d+/', $freeVersion, $match);
```
@triple0t I see this resolved (I unresolved this now), but the issue is still present - once the version is bumped to 10.0.0, it will never match and always use a fallback version.

<img width="546" height="505" alt="Image" src="https://github.com/user-attachments/assets/e27362c9-43e4-4efd-a3f4-414a6e23e3cf" />

Related, we have a [rule for AI to avoid regexp](https://github.com/mailpoet/mailpoet/blob/trunk/.cursor/rules/avoid-regex.mdc) unless absolutely necessary, in part due to hidden issues like this. I think a simple `explode('.', $version)` is more readable and less prone to errors.

triple0t replied:
Fixed in https://github.com/mailpoet/mailpoet/pull/6408/commits/73427af9799af726ffb8c96436b010b62c988fcc

NeosinneR reviewed (commented):
Nice work, thanks @triple0t. Looks good overall, but I do have a question/suggerstion below:

triple0t committed:
Refactor version compatibility check in Updater class

Updated the isVersionCompatible method to use explode for extracting major and minor version numbers from both premium and free versions, improving clarity and maintainability. Adjusted comments for better understanding of the logic.

Aschepikov reviewed (approved):
GJ!
